### ISSUE TITLE
To determine `GAP_SO` sage.env looks for `libgap.so` but it should look for `libgap.so*`

### ISSUE THREAD

**Issue by tornaria** [sagemath/sage#33446]
<div id="comment:0"></div>

My system (void linux) only has

```
$ ls -l /usr/lib/libgap.so*
lrwxrwxrwx 1 root root      15 Jan 14 11:14 /usr/lib/libgap.so.0 -> libgap.so.0.0.0
-rwxr-xr-x 1 root root 2080920 Jan 14 11:14 /usr/lib/libgap.so.0.0.0
```
installed. Indeed, the symlink `/usr/lib/libgap.so -> libgap.so.0.0.0` is shipped in `gap-devel` which is not needed to run sagemath at all.

This is because the implementation of the function `sage.env._get_shared_lib_path()` will look for `libgap.so` instead of `libgap.so*`. The fix is trivial:

```
--- a/src/sage/env.py
+++ b/src/sage/env.py
@@ -293,7 +293,7 @@ def _get_shared_lib_path(*libnames: str) -> Optional[str]:
             if sys.platform == 'darwin':
                 ext = 'dylib'
             else:
-                ext = 'so'
+                ext = 'so*'
 
             if SAGE_LOCAL:
                 search_directories.append(Path(SAGE_LOCAL) / 'lib')
```

Note that this function is only used once in `sage.env` to set `GAP_SO`.

Without the fix above my `GAP_SO` is set to `None` and I get failures like

```
sage: libgap.eval("2+2")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [1], in <module>
----> 1 libgap.eval("2+2")
...
TypeError: expected str, NoneType found
```

Depends on #34391

CC:  @dimpase @orlitzky @williamstein

Component: **packages: standard**

Author: **Gonzalo Tornaría**

Branch/commit - see the corr. PR

Reviewer: **Dima Pasechnik**

_Issue created by migration from https://trac.sagemath.org/ticket/33446_

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:47"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/9feec282d039ca25bbefa50710f8217b775ce111"><code>9feec28</code></a></td><td><code>gap: do not directly dlopen() the gap library</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/704a7953b97ab726e2af610724726aa562bc8bf8"><code>704a795</code></a></td><td><code>singular: do not directly dlopen() the singular library</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/d7145c14ef58acfbe00d8f941d2802e96ad4ba15"><code>d7145c1</code></a></td><td><code>singular: remove LIBSINGULAR_PATH, no longer needed</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/92e5a211c792f86f5325d601abfddf667da6a776"><code>92e5a21</code></a></td><td><code>Revert "src/sage/interfaces/gap_workspace.py: Use hash of GAP_SO to disambiguate the workspace file, not SAGE_LOCAL"</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/b45e555b5711ae10d369b568333940c2aa771053"><code>b45e555</code></a></td><td><code>gap: remove GAP_SO, no longer needed</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/31e3fc4ec8b8687bccd22d2e3161c86cf5553e06"><code>31e3fc4</code></a></td><td><code>gap_workspace_file: include hostname and gap version</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`8ea6ac3`](https://github.com/sagemath/sagetrac-mirror/commit/8ea6ac31fb1a14493618c8cbec17eb3a301870e5)** to **[`31e3fc4`](https://github.com/sagemath/sagetrac-mirror/commit/31e3fc4ec8b8687bccd22d2e3161c86cf5553e06)**

**Comment by tornaria** [comment]
<div id="comment:48" align="right">comment:48</div>

I rebased everything on 9.8.beta1.

I am still unsure what to do about `gap_workspace_file()` and whether the last commit is ok.

I think it would be better to use stuff from `KERNEL_INFO()` as mentioned in [comment:40](#comment%3A40). The problem with that is that `gap_workspace_file()` is called on sage startup and initializing gap is slow. The solution could be to delay calling `gap_workspace_file()` until it is actually used (at which point gap or libgap will be initialized anyway).

For the short run, what is here seems better (with or without the last commit). Right now I still have to patch stuff for void linux since we don't ship `/usr/lib/libgap.so` with runtime packages (and it is a pain since everything will work ok if `gap-devel` happens to be installed, which is the case when building, so the failure tends to be missed).

**Comment by dimpase** [comment]
<div id="comment:49" align="right">comment:49</div>

Could you rebase this on GAP 4.12 - the update has been positively reviewed in #34391 ?

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:50"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. Last 10 new commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/72d57e09637ae74e8d4bf74d82fddedd20853447"><code>72d57e0</code></a></td><td><code>don't use SAGE_LOCAL/share/gap, switch to GAP's "make install"</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/16c2ccb7fcbb95c1fb7f7ce0b0e4b9aec9031497"><code>16c2ccb</code></a></td><td><code>Search for files in both lib/gap and share/gap</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/3abfbd014f80dbda16db0f99d6c81486a1f4cf01"><code>3abfbd0</code></a></td><td><code>Create lib/gap/bin on install</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/73eaa6a0b049b7379fcb557d8c72a26a1d2b243c"><code>73eaa6a</code></a></td><td><code>gap: do not directly dlopen() the gap library</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/8a6cfc43bab9f497d39e8af8d0812ed45f4d094a"><code>8a6cfc4</code></a></td><td><code>singular: do not directly dlopen() the singular library</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7c0f6676ccfe3ba9097647382098c209154de5ef"><code>7c0f667</code></a></td><td><code>singular: remove LIBSINGULAR_PATH, no longer needed</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/fb526096dcf9badd73fac5e8f43d7837f2d2a991"><code>fb52609</code></a></td><td><code>Revert "src/sage/interfaces/gap_workspace.py: Use hash of GAP_SO to disambiguate the workspace file, not SAGE_LOCAL"</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/f1d189cee9c955e92751d84509298316f7118cba"><code>f1d189c</code></a></td><td><code>gap: remove GAP_SO, no longer needed</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/46e48d678d25d6977cba100c220d1997282721d1"><code>46e48d6</code></a></td><td><code>gap_workspace_file: include hostname and gap version</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/1e5dd37e2f6747832107f32833a22ba6e6c38dff"><code>1e5dd37</code></a></td><td><code>gap_workspace_file: use GAP_LIB_DIR and GAP_SHARE_DIR</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`31e3fc4`](https://github.com/sagemath/sagetrac-mirror/commit/31e3fc4ec8b8687bccd22d2e3161c86cf5553e06)** to **[`1e5dd37`](https://github.com/sagemath/sagetrac-mirror/commit/1e5dd37e2f6747832107f32833a22ba6e6c38dff)**

**Comment by tornaria** [comment]
<div id="comment:51" align="right">comment:51</div>

Rebased on top of #34391. I added a commit so `sysinfo.gap` is searched both in `GAP_LIB_DIR` and in `GAP_SHARE_DIR`. Also: use the content of those variables instead of `SAGE_LOCAL`.

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:53"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. Last 10 new commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/93c1becb5f3eba41b0cea43912c2482389320bf9"><code>93c1bec</code></a></td><td><code>ensure make gap-clean still works</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/71c9fe337455d52637e0272fccdd633d8e96fa2e"><code>71c9fe3</code></a></td><td><code>with SAGE_CHECK=yes, one needs io installed to prevent hangs</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/29383be9dbea4e49004ed57d5f23e34155174548"><code>29383be</code></a></td><td><code>backport GAP PR 5235</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/b2a9b2eb6c7948d8cc872ec09746ae48a2cd27b8"><code>b2a9b2e</code></a></td><td><code>gap: do not directly dlopen() the gap library</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/564c5cae572d23e5d654001163889c506093dae0"><code>564c5ca</code></a></td><td><code>singular: do not directly dlopen() the singular library</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/6e1abd31b706a8e4a144d40dc732cf6ce659e6a8"><code>6e1abd3</code></a></td><td><code>singular: remove LIBSINGULAR_PATH, no longer needed</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/ec2476b065ce4cf86eb99017a8ae0dfd4439d3c0"><code>ec2476b</code></a></td><td><code>Revert "src/sage/interfaces/gap_workspace.py: Use hash of GAP_SO to disambiguate the workspace file, not SAGE_LOCAL"</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/eec410bc5376b4bc993a9e4e4bcb035b6a37953b"><code>eec410b</code></a></td><td><code>gap: remove GAP_SO, no longer needed</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/4768b9927eff2ed7e6e6f38e250fee6aae4433be"><code>4768b99</code></a></td><td><code>gap_workspace_file: include hostname and gap version</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/126b0de2e1a1ec137497e433890604789076d2fe"><code>126b0de</code></a></td><td><code>gap_workspace_file: use GAP_LIB_DIR and GAP_SHARE_DIR</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`1e5dd37`](https://github.com/sagemath/sagetrac-mirror/commit/1e5dd37e2f6747832107f32833a22ba6e6c38dff)** to **[`126b0de`](https://github.com/sagemath/sagetrac-mirror/commit/126b0de2e1a1ec137497e433890604789076d2fe)**

**Comment by dimpase** [comment]
Dependencies: **#34391**

**Comment by dimpase** [comment]
<div id="comment:55" align="right">comment:55</div>

let's get #34391 in first - it's a bit hard to review now.

**Comment by tornaria** [comment]
<div id="comment:56" align="right">comment:56</div>

Replying to [Dima Pasechnik](#comment%3A55):
> let's get #34391 in first - it's a bit hard to review now.

Sure, but just for the record: the head of #34391 is
- ​29383be	backport GAP PR 5235
and my branch here is linearly rebased on top of that.

I could squash some of the commits together if that makes it easier to review (ec2476b + 4768b99 + 126b0de are part of the same change, in a sense). That would leave 5 clearly defined commits.

**Comment by dimpase** [comment]
Reviewer: **Dima Pasechnik**

**Comment by dimpase** [comment]
<div id="comment:57" align="right">comment:57</div>

looks good - also tested on macOS

**Comment by vbraun** [comment]
The dependency hasn't been merged yet, so please create a github PR for this ticket
